Fix removing master from replset.

omron93 opened the pull request:
This PR aims to fix 'stepping down server from PRIMARY role'.

When member is going to be removed from replset and this member is PRIMARY, it should check that this won't break replset - that replset is able to vote a new PRIMARY.
In current situation while removing PRIMARY it could happen that not all members of replset are up-to-date and replset user can lost some information.

Code which is now running after post deploy hook successfully initialize replset performs removing PRIMARY from replset. So removing this code from `init-replset.sh` and post-hook kills main scripts and this calls `cleanup()` (there the PRIMARY member is clearly removed).

Also this fixes https://bugzilla.redhat.com/show_bug.cgi?id=1260586

@php-coder @rhcarvalho @bparees Please take a look and review.

EDIT: Updated info after rebase to latest master.

rhcarvalho commented on 3.2/root/usr/bin/run-mongod-replication:
```diff
@@ -34,6 +34,19 @@ function unset_env_vars() {
+
+    # Wait till the new PRIMARY member is elected
+    echo "=> Waiting for the new PRIMARY to be elected ..."
+    mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --quiet --host ${mongo_node} --eval "var done=false;while(done==false){var members=rs.status().members;for(i=0;i<members.length;i++){if(members[i].stateStr=='PRIMARY' && members[i].name!='$(mongo_addr)'){done=true}};sleep(500)};" &>/dev/null
```
`mongo_node` is a variable defined in `replset_supervisor.sh`, undefined here.

omron93 replied:
Thanks. Fixed.

rhcarvalho commented on 3.2/root/usr/bin/run-mongod-replication:
```diff
@@ -34,6 +34,19 @@ function unset_env_vars() {
+
+    # Wait till the new PRIMARY member is elected
+    echo "=> Waiting for the new PRIMARY to be elected ..."
+    mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --quiet --host ${mongo_node} --eval "var done=false;while(done==false){var members=rs.status().members;for(i=0;i<members.length;i++){if(members[i].stateStr=='PRIMARY' && members[i].name!='$(mongo_addr)'){done=true}};sleep(500)};" &>/dev/null
```
I see this is just moving a chunk of code from one file to another. In any case, wouldn't it be nice to split the line to make the oneliner javascript code easier to read / more visible?

omron93 replied:
I agree.

@rhcarvalho   Can I do it in this PR? :-)

rhcarvalho replied:
Would be a welcome improvement!

omron93 replied:
Done. (bun only for code changed in this PR)

rhcarvalho commented on 3.2/root/usr/share/container-scripts/mongodb/replset_supervisor.sh:
```diff
@@ -53,25 +55,12 @@ if [[ "$1" == "initiate" ]]; then
   echo "=> Successfully initialized replSet"
 
+  # Exit this pod
+  kill ${main_process_id}
```
Hmm, sorry I didn't follow. Why are we going to kill `run-mongod-replication`? How does this relate to https://bugzilla.redhat.com/show_bug.cgi?id=1260586?

(BTW, if this PR intends to fix that bug, please add a link in the PR's description)

omron93 replied:
We kill run-mongod-replication to terminate post-hook pod after it initiates replset and creates users. After we kill this pod, `cleanup()` is processed and this pod removed from replset (in master it is removed in `replset_supervisor.sh`)

I was not aware of this bug when I was preparing this PR. I wanted to fix general situation when removing PRIMARY from replset.

And the bug is about this as I understand. In my testing this fixes it.

rhcarvalho reviewed (commented):
@omron93 please clarify what this is trying to solve and how it solves it in the description.

There might be some future here, but I didn't understand yet how this solves the problem, thus asking for clarification.

omron93 commented:
> @omron93 please clarify what this is trying to solve and how it solves it in the description.

https://github.com/sclorg/mongodb-container/pull/186#discussion_r80470367 helps?

> There might be some future here, but I didn't understand yet how this solves the problem, thus asking for clarification.

By " mainly wanted to show possible future improvements." in PR description I wanted to make clear that this PR is not ideal.

For example 

```
echo "=> Waiting for replication to finish ..."
  # TODO: Replace this with polling or a Mongo script that will check if all
  #       members of the cluster are now properly replicated (user accounts are
  #       created on all members).
  sleep 10
```

should be also removed from `replset_supervisor.sh` and waiting in `cleanup()` should be updated to handle this. But I did not want to make this PR more complicated.

(also I wanted to show that after moving code from `replset_supervisor.sh` to `cleanup()` in this PR code of `replset_supervisor.sh` has a similar structure to `replset_petset_supervisor.sh` in #184 )

rhcarvalho commented:
> #186 (comment) helps?

When people (think we in 6 months from now) look at a PR, we shouldn't require them to go through each comment and follow links to understand what's going on.

It is much better when the PR title gives a good short description of what it is about, and then the PR description explains the motivation, why we are doing and what we are doing. Also link there relevant information, such as related GitHub issues and bugzillas.

That's why I asked for clarification in the description, not in a comment or a link.

omron93 commented:
> That's why I asked for clarification in the description, not in a comment or a link.

It is updated.

Commit:
Fix removing master from replset.

JS code reformatting.

omron93 commented:
Rebased and added function to wait that at least one SECONDARY is synced before PRIMARY removing.

@rhcarvalho @php-coder @bparees Any comments please?

php-coder commented:
@omron93 Don't wait my feedback, please. At this time I'm in an another team and working on a different area of the OpenShift, so I don't have time to watch for MongoDB and other images as I did before.

rhcarvalho commented:
@omron93, my situation is similar to @php-coder.

bparees commented:
[test]

bparees commented:
@omron93 lgtm.

omron93 commented:
@php-coder @rhcarvalho Thanks for your work.

Anyone doing the same job instead of you?

bparees commented:
@omron93 we haven't identified a specific owner within the Developer Experience team for the mongo petset work that @php-coder and @rhcarvalho did, but to be honest I think the expectation at this point is for the SCL team to own this sort of thing.

omron93 commented:
> but to be honest I think the expectation at this point is for the SCL team to own this sort of thing.

@bparees I understand. I mainly wanted to know whom to ask for OpenShift related code and code reviews?

hhorak commented on 3.2/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -130,6 +130,30 @@ function replset_addr() {
+    --eval "var i = ${MAX_ATTEMPTS};
+    while(i > 0) {
+      var status=rs.status();
+      var primary_optime=rs.status().members.filter(function(el) {return el.state ==1})[0].optime;
```
Any reason to not use `status` variable instead of calling `rs.status()` again?

hhorak replied:
Calling `rs.status()` more times seems to me like opening gate to race condition, but I might miss something.

omron93 replied:
@hhorak You are right. I overlooked this.

Fixed.

hhorak commented on 3.2/root/usr/share/container-scripts/mongodb/common.sh:
```diff
@@ -130,6 +130,30 @@ function replset_addr() {
+    while(i > 0) {
+      var status=rs.status();
+      var primary_optime=rs.status().members.filter(function(el) {return el.state ==1})[0].optime;
+      if(rs.status().members.filter(function(el) {return el.optime.ts.tojson() == primary_optime.ts.tojson()}).length > 2)
```
I read `...length > 2` like we need to find the master and one another who have the same `optime`, right? It would be better to add comment with short explanation, so reader does not need to figure out it..

omron93 replied:
There should be `length >= 2` (PRIMARY and at least on SECONDARY ~ `>= 2`)
Comment added.

hhorak commented on 3.2/root/usr/bin/run-mongod-replication:
```diff
@@ -28,6 +28,32 @@ function usage() {
+            done=true
+          }
+        };
+        sleep(500)
```
This will make the process sleep 0.5s even after new primary is elected, I'd suggest to not sleep in this case any more, to avoid possible race condition which could potentially happen during that period (that the original master would be re-elected during that time). Many things might happen during 500ms in DB world, so I'd rather be super-paranoid :)

omron93 replied:
Right. Fixed.

Also removed infinite loop and added `MAX_ATTEMPTS` and `SLEEP_TIME`.

hhorak commented:
Just few formal comments in line, otherwise it all makes sense to me.

bparees commented:
> @bparees I understand. I mainly wanted to know whom to ask for OpenShift related code and code reviews?

for now just tag me, when i get overwhelmed i'll find another victim on the devex team :)

hhorak commented:
LGTM.
